client.go

```diff
@@ -410,6 +416,9 @@ func (c *APIClient) RunRawRequestWithHeaders(method, url string, payloadBuffer i
 
 // runRawRequestWithHeaders actually performs the REST calls but allowing custom headers
 func (c *APIClient) runRawRequestWithHeaders(method, url string, payloadBuffer io.ReadSeeker, contentType string, customHeaders map[string]string) (*http.Response, error) {
+	c.mu.Lock()
```

Comment:
Sorry, one more thing while you're at it. Can you lock immediately before the `c.HTTPClient.Do(req)` call, then explicitly unlock right away after? That way we are limiting that window to as small as possible so nothing is blocking while waiting for non-critical things to be done.